The option `--follow-symlinks` can't follow relative symlinks

## Description

The subcommands `lint` and `format` have the option `--follow-symlinks`. However, although it can follow absolute symlinks, it can't follow relative symlinks.

## Steps to reproduce

```console
$ mkdir symlink-test
$ cd symlink-test/
$ echo 'let foo:Int=0' > .foo.swift
$ ln -s .foo.swift bar.swift
$ swift format dump-configuration > .swift-format
$ cd ..
$ swift format lint -rs --follow-symlinks symlink-test/
<unknown>: error: Unable to open .foo.swift: file is not readable or does not exist
```

If the current working directory is the symlink's directory, it works.

```console
$ cd symlink-test
$ swift format lint -rs --follow-symlinks .
.foo.swift:1:9: warning: [Spacing] add 1 space
.foo.swift:1:12: warning: [Spacing] add 1 space
.foo.swift:1:13: warning: [Spacing] add 1 space
```

It's because the following line is using `relativeTo: workingDirectory`.

https://github.com/swiftlang/swift-format/blob/8068ff75e23084bfdc0706a70fa1bc926c9ca71e/Sources/SwiftFormat/Utilities/FileIterator.swift#L152

## Environment

- swift-format (bundled in Swift 6.0.1 toolchain)
- Debian 12 (aarch64)

It seemed that `testFollowSymlinks` failed in macOS. I'll investigate it.

https://ci.swift.org/job/swift-syntax-PR-macOS/4601/console

```
Test Case '-[SwiftFormatTests.FileIteratorTests testFollowSymlinks]' started.
/Users/ec2-user/jenkins/workspace/swift-syntax-PR-macOS/branch-main/swift-format/Tests/SwiftFormatTests/Utilities/FileIteratorTests.swift:37: error: -[SwiftFormatTests.FileIteratorTests testFollowSymlinks] : XCTAssertEqual failed: ("4") is not equal to ("3")
Test Case '-[SwiftFormatTests.FileIteratorTests testFollowSymlinks]' failed (0.699 seconds).
```

--- a/Sources/SwiftFormat/Utilities/FileIterator.swift
+++ b/Sources/SwiftFormat/Utilities/FileIterator.swift
@@ -97,12 +97,12 @@ public struct FileIterator: Sequence, IteratorProtocol {
           output = next
         }
       }
-      if let out = output, visited.contains(out.absoluteURL.standardized.standardizedFileURL.path) {
+      if let out = output, visited.contains(out.standardizedFileURL.path) {
         output = nil
       }
     }
     if let out = output {
-      visited.insert(out.absoluteURL.standardized.standardizedFileURL.path)
+      visited.insert(out.standardizedFileURL.path)
     }
     return output
   }

--- a/Sources/SwiftFormat/Utilities/FileIterator.swift
+++ b/Sources/SwiftFormat/Utilities/FileIterator.swift
@@ -135,7 +135,7 @@ public struct FileIterator: Sequence, IteratorProtocol {
         else {
           break
         }
-        path = URL(fileURLWithPath: destination, isDirectory: false, relativeTo: item).path
+        path = URL(fileURLWithPath: destination, relativeTo: item).path
         fallthrough
 
       case .typeRegular:

--- a/Sources/SwiftFormat/Utilities/FileIterator.swift
+++ b/Sources/SwiftFormat/Utilities/FileIterator.swift
@@ -97,12 +97,12 @@ public struct FileIterator: Sequence, IteratorProtocol {
           output = next
         }
       }
-      if let out = output, visited.contains(out.absoluteURL.standardized.path) {
+      if let out = output, visited.contains(out.absoluteURL.standardized.standardizedFileURL.path) {
         output = nil
       }
     }
     if let out = output {
-      visited.insert(out.absoluteURL.standardized.path)
+      visited.insert(out.absoluteURL.standardized.standardizedFileURL.path)
     }
     return output
   }

--- a/Sources/SwiftFormat/Utilities/FileIterator.swift
+++ b/Sources/SwiftFormat/Utilities/FileIterator.swift
@@ -79,7 +79,7 @@ public struct FileIterator: Sequence, IteratorProtocol {
           else {
             break
           }
-          next = URL(fileURLWithPath: destination)
+          next = URL(fileURLWithPath: destination, relativeTo: next)
           fallthrough
 
         case .typeDirectory:
@@ -135,7 +135,7 @@ public struct FileIterator: Sequence, IteratorProtocol {
         else {
           break
         }
-        path = destination
+        path = URL(fileURLWithPath: destination, isDirectory: false, relativeTo: item).path
         fallthrough
 
       case .typeRegular:

--- a/Tests/SwiftFormatTests/Utilities/FileIteratorTests.swift
+++ b/Tests/SwiftFormatTests/Utilities/FileIteratorTests.swift
@@ -18,6 +18,7 @@ final class FileIteratorTests: XCTestCase {
     try touch("project/.hidden.swift")
     try touch("project/.build/generated.swift")
     try symlink("project/link.swift", to: "project/.hidden.swift")
+    try symlink("project/rellink.swift", relativeTo: ".hidden.swift")
   }
 
   override func tearDownWithError() throws {
@@ -55,7 +56,10 @@ final class FileIteratorTests: XCTestCase {
     // passed to the iterator. This is meant to avoid situations where a symlink could be hidden by
     // shell expansion; for example, if the user writes `swift-format --no-follow-symlinks *`, if
     // the current directory contains a symlink, they would probably *not* expect it to be followed.
-    let seen = allFilesSeen(iteratingOver: [tmpURL("project/link.swift")], followSymlinks: false)
+    let seen = allFilesSeen(
+      iteratingOver: [tmpURL("project/link.swift"), tmpURL("project/rellink.swift")],
+      followSymlinks: false
+    )
     XCTAssertTrue(seen.isEmpty)
   }
 }
@@ -81,11 +85,19 @@ extension FileIteratorTests {
     }
   }
 
-  /// Create a symlink between files or directories in the test's temporary space.
+  /// Create a absolute symlink between files or directories in the test's temporary space.
   private func symlink(_ source: String, to target: String) throws {
     try FileManager.default.createSymbolicLink(
       at: tmpURL(source),
       withDestinationURL: tmpURL(target)
+    )
+  }
+
+  /// Create a relative symlink between files or directories in the test's temporary space.
+  private func symlink(_ source: String, relativeTo target: String) throws {
+    try FileManager.default.createSymbolicLink(
+      atPath: tmpURL(source).path,
+      withDestinationPath: target
     )
   }
 

